From axboe@kernel.dk  Fri Aug  7 15:18:24 2020
From: Jens Axboe <axboe@kernel.dk>
Date: Wed, 5 Aug 2020 13:10:25 -0600
Subject: io_uring: Fix use-after-free in io_sq_wq_submit_work()
To: stable@vger.kernel.org
Message-ID: <69e11137-dd04-5c95-e73c-6c826196d46d@kernel.dk>

From: Guoyu Huang <hgy5945@gmail.com>

when ctx->sqo_mm is zero, io_sq_wq_submit_work() frees 'req'
without deleting it from 'task_list'. After that, 'req' is
accessed in io_ring_ctx_wait_and_kill() which lead to
a use-after-free.

Signed-off-by: Guoyu Huang <hgy5945@gmail.com>
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 fs/io_uring.c |    1 +
 1 file changed, 1 insertion(+)

--- a/fs/io_uring.c
+++ b/fs/io_uring.c
@@ -2232,6 +2232,7 @@ restart:
 		if (io_req_needs_user(req) && !cur_mm) {
 			if (!mmget_not_zero(ctx->sqo_mm)) {
 				ret = -EFAULT;
+				goto end_req;
 			} else {
 				cur_mm = ctx->sqo_mm;
 				use_mm(cur_mm);
